### PR TITLE
Add query params

### PR DESCRIPTION
**Pull request by alex-schneider**

### CODE REVIEW

File: docs/README.md [hunk 1]
```diff
@@ -274,6 +274,39 @@ Endpoints define the entry points of Couper. The mandatory *label* defines the p
...
 |[**`access_control`**](#access_control_attribute)|sets predefined `access_control` for `endpoint`|
 |[**`backend`**](#backend_block) block |configures connection to a local/remote backend service for `endpoint`|
+|[**`remove_query_params`**](#query_params)|<ul><li>a list of query parameters to be removed from the upstream request URL</li></ul> |
+|[**`set_query_params`**](#query_params)|<ul><li>key/value(s) pairs to set query parameters in the upstream request URL</li></ul> |
+|[**`add_query_params`**](#query_params)|<ul><li>key/value(s) pairs to add query parameters to the upstream request URL</li></ul> |
+
+#### Query parameter <a name="query_params"></a>
+
+Couper offers three methods to manipulate the query parameter. The methods are
+executed in the following order:
```
Comment (malud) [inline docs/README.md:284]: We should mention that this is unrelated to the order within the configuration file.

File: handler/proxy.go [hunk 2]
```diff
@@ -319,6 +324,19 @@ func (p *Proxy) Director(req *http.Request) error {
 		if p := getAttribute(evalContext, "path", content); p != "" {
 			path = p
 		}
+		if del, ok := content.Attributes["remove_query_params"]; ok {
```
Comment (malud) [inline handler/proxy.go:327]: "Collecting" and setting the query params should be extracted to a method.

The director method is actually a setup method to prepare the correct dial parameters.

Please refer to `SetRoundtripContext` to make this kind of modifications.

File: handler/proxy.go [hunk 3]
```diff
@@ -330,6 +348,26 @@ func (p *Proxy) Director(req *http.Request) error {
 	req.URL.Scheme = originURL.Scheme
 	req.Host = originURL.Host
 
+	query := req.URL.Query()
```
Comment (malud) [inline handler/proxy.go:351]: This block should be a) extracted to own method as already mentioned and b) splits the "read / logic flow" since the if condition below is related to set the request host above.

File: handler/proxy.go [hunk 4]
```diff
@@ -611,6 +649,26 @@ func setHeaderFields(header http.Header, options OptionsMap) {
 	}
 }
 
+func collectQueryParams(ctx *hcl.EvalContext, attr *hcl.Attribute, dest map[string][]string) {
```
Comment (malud) [inline handler/proxy.go:652]: This method seems to be obsolete at all. Additionally the error handling could be improved.

Please refer to `NewCtxOptions` which returns the required type already with proper hcl diagnostics handling.

**Review by malud — CHANGES_REQUESTED** [review]

File: config/runtime/server.go [hunk 5]
```diff
@@ -420,6 +420,7 @@ func newInlineBackend(
 		if diags.HasErrors() {
 			return nil, diags
 		}
+		bodies = append(bodies, inlineDef.Body())
```
Comment (malud) [inline config/runtime/server.go:423]: Can you please explain the reason why this must be added here? `backendConf` gets created from `inlineBlock` which refers to `inlineDef` and gets added to `bodies` one line below.

Comment (alex-schneider) [reply config/runtime/server.go:423]: The `inlineDef` is the `endpoint` itself. The `inlineBlock` is the `backend` block inside the `endpoint`. This block refers to the `definitions.backend` block. So we have the body from the `definitions.backend` and we add the body from the `backend` block inside the `endpoint` in the next line, but the body of the `endpoint` is missing.

If you remove this line, the test fails, because all of `ae_*` params are missing.

Comment (malud) [reply config/runtime/server.go:438]: I have rebuild the setup to return the `err` instead of fatal log. The failed tests are related to this line. If I remove this one, just the add query params are failing.

File: docs/README.md [hunk 6]
```diff
@@ -274,6 +274,51 @@ Endpoints define the entry points of Couper. The mandatory *label* defines the p
...
 |[**`access_control`**](#access_control_attribute)|sets predefined `access_control` for `endpoint`|
 |[**`backend`**](#backend_block) block |configures connection to a local/remote backend service for `endpoint`|
+|[**`remove_query_params`**](#query_params)|<ul><li>a list of query parameters to be removed from the upstream request URL</li></ul> |
+|[**`set_query_params`**](#query_params)|<ul><li>key/value(s) pairs to set query parameters in the upstream request URL</li></ul> |
+|[**`add_query_params`**](#query_params)|<ul><li>key/value(s) pairs to add query parameters to the upstream request URL</li></ul> |
+
+#### Query parameter <a name="query_params"></a>
+
+Couper offers three methods to manipulate the query parameter. The methods are
+executed in the following order unrelated to the order within the configuration file:
```
Comment (malud) [inline docs/README.md:284]:
```suggestion
The query attributes can be defined unordered within the configuration file but will be executed ordered as follows:
```

File: handler/proxy.go [hunk 7]
```diff
@@ -385,16 +412,35 @@ func (p *Proxy) SetRoundtripContext(req *http.Request, beresp *http.Response) {
...
+		}
+		if add, ok := content.Attributes["add_query_params"]; ok {
+			collectQueryParams(evalCtx, add, addQuery)
+		}
+		if set, ok := content.Attributes["set_query_params"]; ok {
+			collectQueryParams(evalCtx, set, setQuery)
+		}
 	}
 
+	modifyQuery(req.URL, delQuery, setQuery, addQuery)
```
Comment (malud) [inline handler/proxy.go:442]: `req` could be nil here and modify query should not be called for `beresp` context calls.

Comment (alex-schneider) [reply handler/proxy.go:442]: `SetRoundtripContext()` which calls the `modifyQuery()` is never called with `req = nil`.

Comment (malud) [reply handler/proxy.go]: Thats not the point here. This method is called twice. So while reading the beresp we will apply the query params to the outreq again. I have fixed this.

**Review by malud — APPROVED** [review]
